File: algorithms/2_medium/G/01208_get_equal_substrings_within_budget/solution.c
static inline int abs_int(const int a)
{
    return a < 0 ? -a : a;
}

static inline size_t max_usize(const size_t a, const size_t b)
{
    return a > b ? a : b;
}

int equalSubstring(const char* s, const char* t, const int max_cost)
{
    assert(s != NULL && t != NULL && max_cost >= 0);
    size_t lo = 0;
    size_t hi = 0;
    size_t res = 0;
    int cost = 0;

    while (s[hi] != '\0') {
        cost += abs_int((int)s[hi] - (int)t[hi]);
        while (cost > max_cost) {
            cost -= abs_int((int)s[lo] - (int)t[lo]);
            lo += 1;
        }
        res = max_usize(res, hi + 1 - lo);
        hi += 1;
    }

    assert(t[hi] == '\0' && res <= (size_t)INT_MAX);

    return (int)res;
}

File: algorithms/2_medium/G/01208_get_equal_substrings_within_budget/solution.rs
impl Solution {
    pub fn equal_substring(s: String, t: String, max_cost: i32) -> i32 {
        let (s_bytes, t_bytes) = (s.as_bytes(), t.as_bytes());

        assert!(s_bytes.len() == t_bytes.len() && max_cost >= 0);

        let len = s_bytes.len();
        let mut lo = 0_usize;
        let mut res = 0_usize;
        let mut cost = 0_i32;

        for hi in 0..len {
            cost += (s_bytes[hi] as i32 - t_bytes[hi] as i32).abs();
            while cost > max_cost {
                cost -= (s_bytes[lo] as i32 - t_bytes[lo] as i32).abs();
                lo += 1;
            }
            res = res.max(hi + 1 - lo);
        }

        res.try_into().unwrap()
    }
}
